core_lib/src/structure/object.cpp

```diff
@@ -272,38 +274,90 @@ void Object::deleteLayer(Layer* layer)
     }
 }
 
-ColourRef Object::getColour(int i) const
+ColourRef Object::getColour(int index) const
 {
     ColourRef result(Qt::white, "error");
-    if (i > -1 && i < mPalette.size())
+    if (index > -1 && index < mPalette.size())
     {
-        result = mPalette.at(i);
+        result = mPalette.at(index);
+    } else {
+        return mFrontColor;
     }
     return result;
 }
 
 void Object::setColour(int index, QColor newColour)
 {
     Q_ASSERT(index >= 0);
+
+    // To allow a clean color palette, we return if it's empty
+    if (mPalette.isEmpty())
+    {
+        return;
+    }
     mPalette[index].colour = newColour;
 }
 
+void Object::setColourRef(int index, ColourRef newColourRef)
+{
+    if (mPalette.isEmpty())
+    {
+        return;
+    }
+    mPalette[index] = newColourRef;
+}
+
 void Object::addColour(QColor colour)
 {
     addColour(ColourRef(colour, "Colour " + QString::number(mPalette.size())));
 }
 
-bool Object::removeColour(int index)
+void Object::addColourAtIndex(int index, ColourRef newColour)
+{
+    mPalette.insert(index, newColour);
+}
+
+bool Object::shouldDeleteColor(int index)
 {
+    bool usesColor = false;
     for (int i = 0; i < getLayerCount(); i++)
     {
         Layer* layer = getLayer(i);
         if (layer->type() == Layer::VECTOR)
         {
             LayerVector* layerVector = (LayerVector*)layer;
-            if (layerVector->usesColour(index)) return false;
+
+            if (layerVector->usesColour(index))
+            {
+                usesColor = true;
+            }
+        }
+    }
+
+    if (usesColor)
+    {
+        QMessageBox msgBox;
+        msgBox.setText(tr("The color you are trying to delete is currently being used by one or multiple strokes, "
+                       "if you wish to delete it anyway, you accept that the stroke(s) will be bound to the next available color"));
+        QPushButton* cancelButton = msgBox.addButton(tr("Cancel"), QMessageBox::RejectRole);
+        QPushButton* removeButton = msgBox.addButton(tr("Remove anyway"), QMessageBox::AcceptRole);
+
+        msgBox.exec();
+        if (msgBox.clickedButton() == cancelButton)
+        {
+            return false;
+        }
+        else if (msgBox.clickedButton() == removeButton)
+        {
+            return true;
         }
```

Comment:
It's probably not a good idea to pop up a message box in the Object, Object should be pure logic object. It breaks the princile of separating UI & logic, and also makes writing unit test harder as message box will block the function call until someone hits the button.

I would recommend to have a `bool isColourInUse()` and a separate `void removeColor()`. And the moving the message box to ColorPaletteWidget.